### PR TITLE
Mgfractal: Make non-fractal terrain optional

### PR DESCRIPTION
**Pull request by paramat**
screenshot_20190722_004505

Mgfractal needs some attention.
First task is adding a mapgen-specific flag to make non-fractal terrain (sea and underground) optional.
Because this is a fractal generator, generating the fractal only will often be desired. Doing so allows exposing the lower structure of the fractal, and allows creating floating island or 'skyblock' worlds.

This was technically already possible, but only in a difficult-for-most way by altering the noise parameters of 'seabed' noise.
This option also disables 'seabed' noise calculations to increase performance.

Non-fractal terrain is enabled by default.

### CODE REVIEW

File: src/mapgen/mapgen_fractal.cpp [hunk 1]
```diff
@@ -1,7 +1,6 @@
 /*
 Minetest
-Copyright (C) 2015-2018 paramat
-Copyright (C) 2015-2018 kwolekr, Ryan Kwolek <kwolekr@minetest.net>
```
Comment (SmallJoker) [inline src/mapgen/mapgen_fractal.cpp]: Why did you remove this credit? Much was changed in this file, but there's surely a concept or even part of the old code left over.

File: src/mapgen/mapgen_fractal.cpp [hunk 2]
```diff
@@ -66,13 +66,17 @@ MapgenFractal::MapgenFractal(MapgenFractalParams *params, EmergeManager *emerge)
 	julia_z          = params->julia_z;
 	julia_w          = params->julia_w;
 
-	//// 2D terrain noise
-	noise_seabed       = new Noise(&params->np_seabed, seed, csize.X, csize.Z);
+	//// 2D noise
+	if (spflags & MGFRACTAL_TERRAIN)
+		noise_seabed   = new Noise(&params->np_seabed, seed, csize.X, csize.Z);
```
Comment (SmallJoker) [inline src/mapgen/mapgen_fractal.cpp]: Space alignment is now pointless. Please trim the spacing.

File: src/mapgen/mapgen_fractal.cpp (hunk 1 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Ok, i wasn't sure anyway.
The entire mapgen was written by me, but indeed much is code copied form hmmmm's work.
Will re-add.

Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Done.

File: src/mapgen/mapgen_fractal.cpp (hunk 2 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Was aligned to a line below, but i agree not necesssary. Done.

**Comment by paramat** [comment]
Tested existing worlds are not broken.

File: src/mapgen/mapgen_fractal.cpp [hunk 3]
```diff
@@ -66,13 +67,17 @@ MapgenFractal::MapgenFractal(MapgenFractalParams *params, EmergeManager *emerge)
 	julia_z          = params->julia_z;
 	julia_w          = params->julia_w;
 
-	//// 2D terrain noise
-	noise_seabed       = new Noise(&params->np_seabed, seed, csize.X, csize.Z);
+	//// 2D noise
+	if (spflags & MGFRACTAL_TERRAIN)
+		noise_seabed = new Noise(&params->np_seabed, seed, csize.X, csize.Z);
```
Comment (SmallJoker) [inline src/mapgen/mapgen_fractal.cpp]: Please initialize `noise_seabed` to `nullptr` if this flag is not set. Safety first.
PS: What do you think about renaming the pointer to `noise_terrain` to be like its setting and flag?

File: src/mapgen/mapgen_fractal.cpp [hunk 4]
```diff
@@ -153,21 +160,25 @@ void MapgenFractalParams::writeParams(Settings *settings) const
...
-		solid_below = true;
-
-	for (s16 y = search_start; y <= search_start + 128; y++) {
-		if (getFractalAtPoint(p.X, y, p.Y)) {  // Fractal node
+	bool solid_below = false; // Fractal node is present below to spawn on
+	u8 air_count = 0; // Consecutive air nodes above a fractal node
+	s16 search_start = 0; // No terrain search start
+
+	// If terrain present, don't start search below terrain or water level
+	if (spflags & MGFRACTAL_TERRAIN) {
```
Comment (SmallJoker) [inline src/mapgen/mapgen_fractal.cpp]: This conditional branch depends on `noise_seabed`, so I'd rather check `if (seabed) {` to ensure it's REALLY only called if the pointer exists.

File: src/mapgen/mapgen_fractal.cpp [hunk 5]
```diff
@@ -391,24 +403,29 @@ s16 MapgenFractal::generateTerrain()
 	s16 stone_surface_max_y = -MAX_MAP_GENERATION_LIMIT;
 	u32 index2d = 0;
 
-	noise_seabed->perlinMap2D(node_min.X, node_min.Z);
+	if (spflags & MGFRACTAL_TERRAIN)
```
Comment (SmallJoker) [inline src/mapgen/mapgen_fractal.cpp]: Also here. I know that `spflags` are supposed to be unmodified during modification, but forgetting the right checks here could cause issues that are hard to find.

File: src/mapgen/mapgen_fractal.cpp [hunk 6]
```diff
@@ -391,24 +403,29 @@ s16 MapgenFractal::generateTerrain()
...
-					} else if (y <= water_level) {
-						vm->m_data[vi] = n_water;
-					} else {
-						vm->m_data[vi] = n_air;
-					}
+				if (vm->m_data[vi].getContent() != CONTENT_IGNORE)
+					continue;
+
+				s16 seabed_height;
+				if (spflags & MGFRACTAL_TERRAIN)
```
Comment (SmallJoker) [inline src/mapgen/mapgen_fractal.cpp]: Same here

File: src/mapgen/mapgen_fractal.cpp (hunk 3 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Ah thanks, i was thinking about your previous advice about nullptr, then forgot.

Will consider, 'seabed' isn't ideal as it actually also creates islands.

File: src/mapgen/mapgen_fractal.cpp (hunk 4 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Ok.

File: src/mapgen/mapgen_fractal.cpp (hunk 5 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Ok.

File: src/mapgen/mapgen_fractal.cpp (hunk 3 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Well meh, changing the noise name could break worlds, so i'd rather not.

Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Done.

File: src/mapgen/mapgen_fractal.cpp (hunk 4 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Done.

File: src/mapgen/mapgen_fractal.cpp (hunk 5 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Done.

File: src/mapgen/mapgen_fractal.cpp (hunk 6 above)
Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Done.

**Comment by paramat** [comment]
Done and retested.

File: src/mapgen/mapgen_fractal.cpp (hunk 3 above)
Comment (SmallJoker) [reply src/mapgen/mapgen_fractal.cpp]: > Well meh, changing the noise name could break worlds, so i'd rather not.

No it won't break worlds. This is only a variable name which might even be stripped out by the compiler. What matters for compatibility are values bound to file data: such as settings, old default values, map data or mod configurations. C++ internal names have zero impact on that.

Comment (paramat) [reply src/mapgen/mapgen_fractal.cpp]: Ah i see, you mean without changing the setting name too. That would not break worlds but would be messy and confusing, for no good reason. All noise parameter setting names match the internal variable names.
This noise is best named 'seabed', but the flag is for the underground too so is best named 'terrain' and not 'seabed'. There's no need for the 2 to match.
